src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -7,400 +7,441 @@
+private[concurrent] final object Promise {
+    private final class Link[T](to: DefaultPromise[T]) extends AtomicReference[DefaultPromise[T]](to) {
 
-  import scala.concurrent.Future
-  import scala.concurrent.impl.Promise.DefaultPromise
+      final def promise(owner: DefaultPromise[T]): DefaultPromise[T] =
+        relink(link = this, target = this.get(), owner = owner)
 
-  override def transform[S](f: Try[T] => Try[S])(implicit executor: ExecutionContext): Future[S] = {
-    val p = new DefaultPromise[S]()
-    onComplete { result => p.complete(try f(result) catch { case NonFatal(t) => Failure(t) }) }
-    p.future
-  }
+      @tailrec private[this] final def root(of: DefaultPromise[T], owner: DefaultPromise[T]): DefaultPromise[T] = {
+        val value = of.get()
+        if (value.isInstanceOf[Link[T]]) root(of = value.asInstanceOf[Link[T]].get(), owner = owner)
+        else if ((owner ne null) && value.isInstanceOf[Try[T]]) { // Link completion detected, unlink.
+          unlink(owner = owner, value = value.asInstanceOf[Try[T]])
+          owner
+        } else of
+      }
+
+      @tailrec final def relink(link: Link[T], target: DefaultPromise[T], owner: DefaultPromise[T]): DefaultPromise[T] = {
+        val current = link.get()
+        val newTarget = root(of = target, owner = owner)
+        if ((current eq newTarget) || (owner eq newTarget) || compareAndSet(current, newTarget)) newTarget
+        else relink(link = link, target = target, owner = owner) // TODO: verify would it be safe to resume with `target = newTarget`?
+      }
 
-  // If possible, link DefaultPromises to avoid space leaks
-  override def transformWith[S](f: Try[T] => Future[S])(implicit executor: ExecutionContext): Future[S] = {
-    val p = new DefaultPromise[S]()
-    onComplete {
-      v => try f(v) match {
-        case fut if fut eq this => p complete v.asInstanceOf[Try[S]]
-        case dp: DefaultPromise[_] => dp.asInstanceOf[DefaultPromise[S]].linkRootOf(p)
-        case fut => p completeWith fut
-      } catch { case NonFatal(t) => p failure t }
+      @tailrec private[this] final def unlink(owner: DefaultPromise[T], value: Try[T]): Unit = {
+        val l = owner.get()
+        if (l.isInstanceOf[Link[T]])
+          unlink(owner = if (owner.compareAndSet(l, value)) l.asInstanceOf[Link[T]].get() else owner, value = value)
+        else if(l.isInstanceOf[Callbacks[T]]) owner.tryComplete(value)
+        // The following left implicit: else /* if (l.isInstanceOf[Try[T]]) */ ()
+      }
     }
-    p.future
-  }
 
-  override def toString: String = value match {
-    case Some(result) => "Future("+result+")"
-    case None => "Future(<not completed>)"
-  }
-}
+    // requireNonNull is paramount to guard against null completions
+    private[this] final def resolve[T](value: Try[T]): Try[T] =
+      if (requireNonNull(value).isInstanceOf[Success[T]]) value
+      else {
+        val t = value.asInstanceOf[Failure[T]].exception
```

Comment:
I still think we should keep `requireNonNull` just in case, but it seems to me that since `null.isInstanceOf[_]` is always false, and the other path immediately invokes `.exception` on `value`, a `null` completion will not get through, and `requireNonNull` is not "paramount"

Reply:
@NthPortal Exactly—we could rely on `.exception` to throw the NPE but it would seem like a bug (and would be brittle). Also requireNonNull can possibly be optimized by theJVM.